Documentation/proposal/fetchers_refactor.md

```diff
@@ -0,0 +1,125 @@
+
+## Fetching logic with distribution
+
+If and when the distribution concept and implementation in #2953 will land, the next step will be a fetchers refactor to fully cover that.
+
+These is the proposed idea:
+
+## Distribution handlers
+* Every distribution will have one or more dedicated handlers:
+
+### Distribution goals:
+
+* The handlers will receive an input distribution string and additional arguments (like the current ability to provide/override a signature with a local file). They will talk with the image store.
+* From the input string they'll implement the distribution specific logic to fetch an image (this means calling one or multiple different transport fetching) and import it in the image store.
+
+
+Examples:
+
+
+**archive distribution**
+
+base archive distribution handler (may choose a sub handler based on the archive format) -> transport handler to fetch the ArchiveURL (from file://, http:// etc...)
+
+**docker distribution**
+
+docker handler ->  The docker handler may choose to use docker2aci and import an ACI in the store or natively call transport handlers to fetch v2/oci blobs with a transport and import them in the store.
+
+**appc distribution**
+
+appc handler -> The appc handler will call appc discovery and fetch the discovered URLs. An interesting thing is that this fetcher should internally convert the discovered ACI URL to a child distribution of type Archive without rewriting all the fetching logic
+
+
+# Distributions handler vs transport handlers
+The above logic will also creates a better separation between the distribution and the transport layers.
+
+For example there may exist multiple transport plugins (file, http, s3, bittorrent etc...) to be called by an archive distribution.
+
+
+
+## Distribution Handlers plugins
+
+With this approach the distributions handler will live in the stage0. If there's the need to make them plugins the main tricky points is that they need to talk with an image store, so it'll be difficult to code them in a different language than go.
```

Comment:
What I wanted in my plugins proposal was to enable support for alternate transports, so I'm fine keeping this logic inside the stage0.